tests/test_2408_layoutbuilder_in_numba.py

```diff
@@ -0,0 +1,1462 @@
+
+    @numba.njit
+    def f30(builder):
+        content = builder.append_invalid()
+        content.append(np.nan)
+
+    @numba.njit
+    def f31(builder, data):
+        content = builder.extend_valid(len(data))
+        content.extend(data)
+
+    @numba.njit
+    def f32(builder, size):
+        content = builder.extend_invalid(size)
+        content.append(np.nan)
+        content.append(np.nan)
+
+    builder = lb.BitMasked(np.uint8, lb.Numpy(np.float64), True, True)
+    assert len(builder) == 0
+
+    f29(builder)
+    f30(builder)
+    data = np.array([3.3, 4.4, 5.5], dtype=np.float64)
+    f31(builder, data)
+
+    f32(builder, 2)
+
+    array = builder.snapshot()
+    assert ak.to_list(array) == [
+        1.1,
+        2.2,
+        3.3,
+        4.4,
+        5.5,
+        6.6,
+        None,
+        3.3,
+        4.4,
+        5.5,
+        None,
+        None,
+    ]
+
+    error = ""
+    assert builder.is_valid(error), error
+
+    assert len(builder) == 12
+    builder.clear()
+    assert len(builder) == 0
+
+
+def test_Unmasked_append_extend():
+    @numba.njit
+    def f33(builder):
+        content = builder._content
+        content.append(1.1)
+        content.append(2.2)
+        content.append(3.3)
+        content.append(4.4)
+        content.append(5.5)
+        content.append(6.6)
+
+    @numba.njit
+    def f34(builder, data):
+        content = builder._content
+        content.extend(data)
+
+    builder = lb.Unmasked(lb.Numpy(np.float64))
+
+    f33(builder)
+    data = np.array([3.3, 4.4, 5.5], dtype=np.float64)
+    f34(builder, data)
+
+    array = builder.snapshot()
+    assert ak.to_list(array) == [
+        1.1,
+        2.2,
+        3.3,
+        4.4,
+        5.5,
+        6.6,
+        3.3,
+        4.4,
+        5.5,
+    ]
+
+    error = ""
+    assert builder.is_valid(error), error
+
+    assert len(builder) == 9
+    builder.clear()
+    assert len(builder) == 0
+
+
+def test_Record_content():
+    @numba.njit
+    def field_index(builder, name):
+        return builder._field_index(name)
+
+    @numba.njit
+    def fill(builder):
+        content_one = builder.content(0)  # "one")
+        content_one.append(1.1)
+        content_two = builder.content(1)  # "two")
+        content_two.append(1)
+        content_three = builder.content(2)  # "three")
+        content_three.append(111)
+
+    builder = lb.Record(
+        [
+            lb.Numpy(np.float64),
+            lb.Numpy(np.int64),
+            lb.Numpy(np.uint8, parameters={"__array__": "char"}),
+        ],
+        ["one", "two", "three"],
+    )
+    assert field_index(builder, "one") == 0
+    assert field_index(builder, "two") == 1
+    assert field_index(builder, "three") == 2
+
+    with pytest.raises(ValueError):
+        field_index(builder, "four")  # ValueError: tuple.index(x): x not in tuple
+
+    fill(builder)
+    array = builder.snapshot()
+    assert ak.to_list(array) == [{"one": 1.1, "three": "o", "two": 1}]  # ???
```

Comment:
Hmm... I looked into the order issue: printing here the `array` from the `snapshot` looks correct and so does the `print(ak.to_list(array))`:
```python
[{'one': 1.1, 'two': 1, 'three': 'o'}]
```

The `assert ak.to_list(array)  == [{'one': 1.1, 'two': 1, 'three': 'o'}]` in Python prompt.

Reply:
I forgot to mention that I tested it with Python 3.9 + Numba 0.56 and Python 3.10 + Numba 0.57. The result is the same: the assert check of `ak.to_list(array)` is sorted by field.

Reply:
It's possible to drill down into this by finding out if the `builder.contents` and `builder.fields` are also out of order. Or maybe after the snapshot, the `array.fields` and `array.contents` of the `RecordArray`. At some point, they get out of order.

The order `"one", "three", "two"` is suggestive of alphabetical sorting, but with only three items, that might be a coincidence. This might be some random hash order that looks alphabetical for these three.